storagemarket/deal_execution.go

```diff
@@ -28,49 +29,44 @@ import (
 	"github.com/libp2p/go-libp2p-core/event"
 )
 
-func (p *Provider) doDeal(deal *types.ProviderDealState) {
+func (p *Provider) doDeal(deal *types.ProviderDealState, dh *dealHandler) {
 	// Set up pubsub for deal updates
-	bus := eventbus.NewBus()
-	pub, err := bus.Emitter(&types.ProviderDealState{}, eventbus.Stateful)
+	pub, err := dh.bus.Emitter(&types.ProviderDealState{}, eventbus.Stateful)
 	if err != nil {
 		err := fmt.Errorf("failed to create event emitter: %w", err)
+		p.cleanupDeal(deal)
```

Comment:
It seems like p.cleanupDeal(deal) gets called on every code path - would it make sense to just have a single `defer p.cleanupDeal(deal)` at the start of the function?

Reply:
@dirkmc It's a bit tricky -> for eg: we need to ensure we've cleaned up a deal before calling `failDeal`. 
Also, we going-ahead we will probably return from this function after the deal has expired/slashed but we wanna clean up the deal state as soon as it's been handed over for sealing.

I tried doing this and the. code looks hard to reason about. Let's keep it this way for now -> it's really easy to see what's happening.

Reply:
Ok